lib/Dialect/Triton/Transforms/RewriteTMAToLegacy.cpp

```diff
@@ -0,0 +1,428 @@
+                    offsets),
+        generateMask(rewriter, op->getLoc(), getBoundaryCheck(op), block_shape,
+                     offsets, shape),
+        generateOther(rewriter, op->getLoc(), getPadding(op), base,
+                      block_shape),
+        getCache(op), getEvict(op), getIsVolatile(op));
+    new_load->setAttrs(filterSegmentSizes(op->getAttrs()));
+
+    return llvm::success();
+  }
+};
+
+struct RewriteStorePattern : OpConversionPattern<triton::DescriptorStoreOp> {
+  using OpConversionPattern<triton::DescriptorStoreOp>::OpConversionPattern;
+
+  llvm::LogicalResult
+  matchAndRewrite(triton::DescriptorStoreOp op, OneToNOpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    const auto block_shape = op.getDesc().getType().getBlockType().getShape();
+    const auto rank = block_shape.size();
+    assert(adaptor.getDesc().size() == 1 + 2 * rank &&
+           "Expected tensor descriptors to be "
+           "broken down into a ptr and "
+           "`rank` shapes and `rank` strides");
+
+    auto base = adaptor.getDesc().front();
+    auto shape = adaptor.getDesc().slice(1, rank);
+    auto strides = adaptor.getDesc().slice(1 + rank, rank);
+    // Note that indices aren't converted so
+    // we can get them directly here
+    auto offsets = castToI64(rewriter, op.getIndices());
+
+    auto new_store = rewriter.replaceOpWithNewOp<triton::StoreOp>(
+        op,
+        generatePtr(rewriter, op->getLoc(), block_shape, base, strides,
+                    offsets),
+        op.getSrc(),
+        generateMask(rewriter, op->getLoc(), getBoundaryCheck(op), block_shape,
+                     offsets, shape),
+        getCache(op), getEvict(op));
+    new_store->setAttrs(filterSegmentSizes(op->getAttrs()));
+
+    return llvm::success();
+  }
+};
+
+/**
+ * @brief This implements the pass for converting triton tensor descriptor
+ * loads/stores into indexed loads/stores.
+ *
+ * The key idea is that each tensor descriptor can be broken down into multiple
+ * values. Suppose we have a tensor pointer with rank r, we can cast that tensor
+ * descriptor value to and from 1+2r values: a tensor pointer value and two i32
+ * value for each dimension representing the dynamic shape and strides.
+ *
+ * As in normal conversion patterns, individual operations can be converted
+ * using casted tensor descriptors and offsets and casting the results back to
+ * tensor pointers.
+ *
+ * We have special handling for TMA loads/stores and the make tensor descriptor
+ * op.
+ *
+ * @note Why use the conversion pattern rewriter? In most cases the defining
+ * operation of a tensor descriptor will be a make tensor descriptor op.
+ * However, this isn't always true - for example, if the tensor descriptor is a
+ * function argument or is in a conditional statement, we need better tracking
+ * of the pointer, shape, and strides.
+ */
+class RewriteTritonTmaToLegacyPass
```

Comment:
NIT: more accurate name
```suggestion
class TritonRewriteTensorDescriptorToPointerPass
```